server/src/main/java/org/elasticsearch/gateway/WriteStateException.java

```diff
@@ -38,4 +38,8 @@ public WriteStateException(boolean dirty, String message, Exception cause) {
     public boolean isDirty() {
         return dirty;
     }
+
+    public void resetDirty() {
+        this.dirty = false;
```

Comment:
This seems bad, and I can't immediately see why this is needed. I think `dirty` should be `final`. I also think we shouldn't have methods on `MetaStateService` that clear the dirty flag like this - I think the caller should handle the dirty flag in any exception thrown appropriately.

Reply:
I agree that this is bad. The reason why this is needed is that we're only concerned about manifest file write that throws WriteStateException with dirty flag set. Other writes that fail with the dirty flag set are not harmful, because they are not reflected in the manifest file.

Reply:
Or probably even better solution would be to wrap WriteStateException in another WriteStateException and changing dirty flag to false for global and index writes. This will also address your comment with logging nicely.

Reply:
Yes wrapping in another exception seems better.